# Calculating Specific Humidity (q) from Relative Humidity (RH)

__THIS IS CURRENTLY NOT CORRECT/TESTED PLEASE DO NOT TRUST THIS__

### This is an attempt to decipher the formulas used to calculate Specific Humidity (SH) from Relative Humidity (RH) while also providing an example for checking code at the end of it.

The following formula and explanation are found here: https://earthscience.stackexchange.com/questions/5076/how-to-calculate-specific-humidity-with-relative-humidity-temperature-and-pres/

> If you are given $RH$ (in the range [0,1]), $T$(K) and $p$(Pa) you can proceed as follows.
>
> Knowing that
> $$RH = \dfrac{e}{e_s},$$
> $$w = \dfrac{e\ R_d}{R_v(p-e)},$$
> and
> $$q = \dfrac{w}{w+1}$$
> 
> Then we can solve for specific humidity $q$
> 
> Rather than combining this into a single formula and solving, it is more straightforward to present this incrementally.
> 
> First, find $e_s(T)$
> where
> $$e_s(T) = e_{s0}\exp\left[\left(\dfrac{L_v(T)}{R_v}\right)\left(\dfrac{1}{T_0}-\dfrac{1}{T}\right)\right]$$
> and then find e from the first formula ($e = RH*e_s$). Then plug $e$ into the formula for $w$ and then that result into the formula for $q$. 
> 
> Variables used:
> - $q$ specific humidity or the mass mixing ratio of water vapor to total air (dimensionless)
> - $w$ mass mixing ratio of water vapor to dry air (dimensionless)
> - $e_s(T)$ saturation vapor pressure (Pa)
> - $e_s0$ saturation vapor pressure at $T_0$ (Pa)
> - $Rd$ specific gas constant for dry air (J kg$^{-1}$ K$^{-1}$)
> - $Rv$ specific gas constant for water vapor (J kg$^{-1}$ K$^{-1}$)
> - $p$ pressure (Pa)
> - $L_v(T)$ specific enthalpy of vaporization (J kg$^{-1}$)
> - $T$ temperature (K)
> - $T0$ reference temperature (typically 273.16 K) (K)
> 
> To display pressure in millibars, conveniently, 1 millibar = 1 hectopascal.

In [1]:
import math

# External Inputs
Temp = 22 # Degrees C
Hum = 60 # in %
Pres = 1005 # mbar

# Constants
T0 = 273.15
es0 = 611.29 # CRC Handbook of Chemistry and Physics by David R Lide
LvT = 2257*1000 # https://en.wikipedia.org/wiki/Enthalpy_of_vaporization then change J/g to J/kg
Rv = 461.5 # https://en.wikipedia.org/wiki/Gas_constant#Specific_gas_constant
Rd = 287.052874 # https://en.wikipedia.org/wiki/Gas_constant#Specific_gas_constant

# External Inputs ready to feed in
T = T0+Temp # Degrees C into Kelvin
p = Pres*100 # mbar into Pascals
RH = 60/100 # Humidity as a decimal not a %


In [2]:
# First part of the formula
esT = es0*(math.exp((LvT/Rv)*((1/T0)-(1/T))))
e = RH*esT
w = (e*Rd)/(Rv*(p-e))
q = w/(w+1)

In [3]:
print(esT)
print(e)
print(w)
print(q)

2321.8807982413255
1393.1284789447952
0.008743344636439247
0.008667561162042096


### Results

With a Temp of 22°C and a Humidity of 60 %RH at 1005 mbar we get,
- $e_{s0}$ = 2321.8807982413255
- $e$ = 1393.1284789447952
- $w$ = 0.008743344636439247
- $q$ = 0.008667561162042096